### DISCUSSION TITLE
TriggerHandler and Service Layer

### DISCUSSION THREAD

**Discussion by reignman5** [apex-enterprise-patterns/fflib-apex-common#341] [Q&A] [answered]
Hey,

trying to integrate the new domain structure and just ran into a problem where I am little confused.
Scenario:
On before update I have a method that if specific conditions are met is doing multiple things. This method contains multiple methods.
```
public class CustomObjectTriggerHandler{

public override onBeforeUpdate(){
doProcess();
}
private void doProcess() {
    doProcess1();
    doProcess2();
    doProcess3();
  }

  private void doProcess1() {
    //do something
  }
  private void doProcess2() {
    //do something
  }
  private void doProcess3() {
    //do something
  }
}
`````
Now I want to call the "doProcess" logic from a button on this custom object regardless of whether the specific conditions are met. Can I call the "doProcess" Method from my service layer or would I refactor this domain logic into a service layer?

Appreciate the feedback

**Comment by wimvelzeboer** [comment, marked as answer]
Hi, @reignman5 
You would normally have a method on your controller which calls a service method to execute business logic.
When that business logic is handling Lists of (S)Objects than that list is warped inside a domain to perform getter, setter and filter operations on that list.

**Reply by reignman5** [reply]
Thanks for quick response.
But if dont call methods on the TriggerHandler wouldnt I duplicate business logic?
Because now I would have the "doProcess" logic on the triggerHandler and also on the service class, as I am trying to do the same from an update and on a button click. At least I would duplicate the sequence of domain method calls

**Reply by wimvelzeboer** [reply]
@reignman5 
The controller and the trigger handler are only points of entry. The actual business logic resides on the service layer (which can utilise domain(s) and selector(s) layers).
Here is an example:

```
public with sharing class MyController()
{
    public void onButtonClick()
    {
        AccountsService.doBusinessProcess();
    }
}
```

```
public with sharing class AccountsTriggerHandler extends fflib_SObjectDomain
{
    ....
    public override void onAfterUpdate(...)
    {
        onChangedStatusDoBusinessProcess();
    }

    private void onChangedStatusDoBusinessProcess()
    {
        List<Account> changedRecords = getChangedRecords(new List<SObjectField>{ Account.Status__c });
        if (changedRecords.isEmpty()) return;
        AccountsService.doBusinessProcess();
    }    
   ....  
} 
```
Here you see that both entry point contain minimal logic and calls the business functionality on the service layer.

**Reply by reignman5** [reply]
@wimvelzeboer 
makes sense. Could you elaborate what logic would be put into the service layer and which logic would be put on the domain layer? I know of #109 , but seems like it is out of date(?).

Also want to thank you again for your help!

**Reply by wimvelzeboer** [reply]
Hi @reignman5,

Let me give you an example in this pseudo code block, where you see service method in more detailed
```
public void doBusinessProcess(Set<Id> accountIds)	
{ 	    	    	  	    												 
    IAccounts accounts = Accounts.newInstance(accountIds);   // create domain with the provided set of account Ids

    IAccounts emeaAccounts = accounts.selectByRegion('EMEA');  // filter the domain on just only accounts from EMEA region
    if (emeaAccounts.isEmpty()) return;

    // create a domain with all the contact records linked to the EMEA accounts
    IContacts contacts = 
            Contacts.newInstance(
                    ContactsSelector.newInstance()
                            .selectByAccountId(
                                    emeaAccounts.getRecordIds()
                            );

    // Let the domain perform a change to the region field on all the contact records
    contacts.setRegion('EMEA');

    fflib_ISObjectUnitOfWork unitOfWork = Application.UnitOfWork.newInstance();

    // Register the records as dirty with the UnitOfWork
    unitOfWork.registerDirty(contacts.getRecords());

    // Perform DML operation
    unitOfWork.commitWork();
}
```

Please also review some of the training materials I created. It also contains a list with "rules" for each layer.
https://github.com/wimvelzeboer/fflib-training-sessions/tree/main/SoC

**Reply by reignman5** [reply]
@wimvelzeboer 
Thanks for that link! Will read through it.
Can I open an issue there if I something is unclear?

**Reply by wimvelzeboer** [reply]
@reignman5 Sure!
